node-template/src/service.rs

```diff
@@ -29,113 +25,107 @@ native_executor_instance!(
-		},
-		LightTransactionPoolApi = transaction_pool::ChainApi<
-			client::Client<LightBackend<Self>, LightExecutor<Self>, Block, RuntimeApi>,
-			Block
-		> {
-			|config, client| Ok(TransactionPool::new(config, transaction_pool::ChainApi::new(client)))
-		},
-		Genesis = GenesisConfig,
-		Configuration = NodeConfig,
-		FullService = FullComponents<Self>
-			{ |config: FactoryFullConfiguration<Self>|
-				FullComponents::<Factory>::new(config)
-			},
-		AuthoritySetup = {
-			|service: Self::FullService| {
-				if service.config().roles.is_authority() {
-					let proposer = ProposerFactory {
-						client: service.client(),
-						transaction_pool: service.transaction_pool(),
-					};
-					let client = service.client();
-					let select_chain = service.select_chain()
-						.ok_or_else(|| ServiceError::SelectChainRequired)?;
-					let aura = start_aura::<_, _, _, _, _, AuraAuthorityPair, _, _, _>(
-						SlotDuration::get_or_compute(&*client)?,
-						client.clone(),
-						select_chain,
-						client,
-						proposer,
-						service.network(),
-						service.config().custom.inherent_data_providers.clone(),
-						service.config().force_authoring,
-						Some(service.keystore()),
-					)?;
-					service.spawn_task(Box::new(aura.select(service.on_exit()).then(|_| Ok(()))));
-				}
 
-				Ok(service)
-			}
-		},
-		LightService = LightComponents<Self>
-			{ |config| <LightComponents<Factory>>::new(config) },
-		FullImportQueue = AuraImportQueue<
-			Self::Block,
-		>
-			{ |
-				config: &mut FactoryFullConfiguration<Self>,
-				client: Arc<FullClient<Self>>,
-				_select_chain: Self::SelectChain,
-				transaction_pool: Option<Arc<TransactionPool<Self::FullTransactionPoolApi>>>,
-			| {
-					import_queue::<_, _, aura_primitives::sr25519::AuthorityPair, _>(
-						SlotDuration::get_or_compute(&*client)?,
-						Box::new(client.clone()),
-						None,
-						None,
-						client,
-						config.custom.inherent_data_providers.clone(),
-						transaction_pool,
-					).map_err(Into::into)
-				}
-			},
-		LightImportQueue = AuraImportQueue<
-			Self::Block,
-		>
-			{ |config: &mut FactoryFullConfiguration<Self>, client: Arc<LightClient<Self>>| {
-					let fprb = Box::new(DummyFinalityProofRequestBuilder::default()) as Box<_>;
-					import_queue::<_, _, AuraAuthorityPair, TransactionPool<Self::FullTransactionPoolApi>>(
-						SlotDuration::get_or_compute(&*client)?,
-						Box::new(client.clone()),
-						None,
-						None,
-						client,
-						config.custom.inherent_data_providers.clone(),
-						None,
-					).map(|q| (q, fprb)).map_err(Into::into)
-				}
-			},
-		SelectChain = LongestChain<FullBackend<Self>, Self::Block>
-			{ |config: &FactoryFullConfiguration<Self>, client: Arc<FullClient<Self>>| {
+/// Starts a `ServiceBuilder` for a full service.
+///
+/// Use this macro if you don't actually need the full service, but just the builder in order to
+/// be able to perform chain operations.
+macro_rules! new_full_start {
```

Comment:
The reason why this macro exists is that we don't want to build a full service when for example importing blocks. In order to import blocks, we only need the client, select-chain and import-queue.

I'm using a macro rather than a function because the return type would be very verbose. This is `node-template` and is supposed to be tweakable for the user, so anything that looks complicated is negative.

Maybe we can introduce a `ServiceIntermediaryType` trait in the future, or something like that. Since the design needs to be figured out, I'd like to delay that for further refactoring, if that's ok.